runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -7040,10 +7040,20 @@ static bool genZeroInitObject2(
    auto headerSize = isArrayNew ? TR::Compiler->om.contiguousArrayHeaderSizeInBytes() : TR::Compiler->om.objectHeaderSizeInBytes();
    // If we are using full refs both contiguous and discontiguous array header have the same size, in which case we must adjust header size
    // slightly so that rep stosb can initialize the size field of zero sized arrays appropriately
-   if (!cg->comp()->target().is32Bit() && !TR::Compiler->om.compressObjectReferences() && isArrayNew)
+   //   #bits per section (compressed refs): | 32 bits |  32 bits   | 32 bits | 32 bits |   32 bits   |   32 bits    |
+   //   zero sized arrays:                   |  class  | mustBeZero |   size  | padding |          dataAddr          |
+   //   smallest contiguous array:           |  class  |    size    |      dataAddr     | 1 byte + padding |  other  |
+   //   In order for us to successfully initialize the size field of a zero sized array in compressed refs
+   //   we must subtract 8 bytes (sizeof(dataAddr)) from header size. And in case of full refs we must
+   //   subtract 16 bytes from the header in order to properly initialize the zero sized field. We can
+   //   accomplish that by simply subtracting the offset of dataAddr field, which is 8 for compressed refs
+   //   and 16 for full refs.
+#if defined(TR_TARGET_64BIT)
+   if (!cg->comp()->target().is32Bit() && isArrayNew)
       {
-      headerSize -= 8;
+      headerSize -= ((TR_J9VMBase *)(cg->fe()))->getOffsetOfContiguousDataAddrField();
```

Comment:
Is the cast needed here?

Reply:
Yes, we get a "no member named ‘getOffsetOfContiguousDataAddrField’ in class TR_FrontEnd" error if we don't cast it to `TR_J9VMBase *`

Reply:
Oh I see, I saw the parenthesis in the wrong place. I thought we were casting whatever `getOffsetOfContiguousDataAddrField` returned to `TR_J9VMBase *`. A change I would like to see made is to make use of `static_cast` here instead of a C style cast as the C++ variant is type safe.

Reply:
Sounds good!